Add Support for GitHub Actions Annotations

cpisciotta opened the pull request:
Resolves: #38

cpisciotta committed:
Add OutputFormat

cpisciotta committed:
Create GitHubActionsLogTests.swift

cpisciotta commented:
Hey, @pepicrft and @fortmarek! Would one of you be able to take a look at this initial implementation? I'm hoping to get some direction to bring this into a more production-ready state. Alternatively, I'd appreciate you forwarding this to someone with context!

I'm seeing some issues around duplicate annotations, and I haven't settled on a general format for the output.

Some initial questions around formatting:
- If reporting for GitHub Actions, should the output be limited to warnings and errors?
- Should unsuccessful steps have markup (color, leading icon, etc.)?
- If including all steps, should successful steps have markup (color, leading icon, etc.)?

pepicrft commented on Sources/XcbeautifyLib/Array+Safe.swift:
```diff
@@ -0,0 +1,15 @@
+//
+//  Created by Charles Pisciotta on 6/19/23.
+//
+
```
I'd remove these comments since it's information already tracked at the git level.

pepicrft commented on Sources/xcbeautify/Xcbeautify.swift:
```diff
@@ -23,6 +23,9 @@ struct Xcbeautify: ParsableCommand {
     @Flag(name: .long, help: "Disable the colored output")
     var disableColoredOutput = (ProcessInfo.processInfo.environment["NO_COLOR"] != nil)
 
+    @Option(help: "Specify an OutputFormat.")
```
The Swift type `OutputFormat` is an implementation detail that's not known when interfacing with the CLI. I'd say something along the lines of:

```
Specify the output format (e.g. `--github-action`)
```

pepicrft commented on Sources/xcbeautify/OutputFormat+ExpressibleByArgument.swift:
```diff
@@ -0,0 +1,12 @@
+//
+//  Created by Charles Pisciotta on 6/15/23.
+//
+
```
I'd remove these comments.

pepicrft commented on Tests/XcbeautifyLibTests/GitHubActionsLogTests.swift:
```diff
@@ -0,0 +1,678 @@
+    }
+
+    func testCheckDependencies() {
+    }
```
Is the plan to add tests for those as well?

cpisciotta replied:
I think this'll need to be a follow up PR because I don't know what the expected output is here. It wasn't defined in the pre-existing implementation (for TerminalRenderer).

pepicrft reviewed (approved):
Thanks a lot @cpisciotta for taking this one on. The approach that you took looks sensible and in line with the coding practices of the repository. I left a few minor comments throughout the PR.

pepicrft commented:
> If reporting for GitHub Actions, should the output be limited to warnings and errors?

That's what I'd advise for now since it's the information developers usually need to diagnose failing builds. If developers need more granularity, we can iterate on your foundation.

> Should unsuccessful steps have markup (color, leading icon, etc.)?

That's up to you to make the call. I'd advise to output the messages in a format that when presented by GitHub someone is able to tell:
- The severity of the message
- The meaning of the message

Beyond that (e.g. icons), I'm not sure it adds that much value.

> If including all steps, should successful steps have markup (color, leading icon, etc.)?

Similarly, I'd keep this one raw for now, and iterate from there. If you have ideas for how to leverage icons or colours to enhance the message, don't hesitate to share the options here and we can discuss what's the best one.

eliperkins commented on Sources/XcbeautifyLib/String+Beautify.swift:
```diff
@@ -313,18 +323,44 @@ extension String {
+                return testSuite
+            case .standard:
+                return _colored ? testSuite.s.Bold : testSuite
+            }
```
Rather than adding these switch statements to each place a string is returned, I wonder if it makes sense to have a `OutputRendering` protocol, and then have a `TerminalRenderer` (which is the previous implementation) and a `GitHubActionsRenderer` (which is this new implementation).

That way, the renderer could then be passed in based on the format, and perhaps this could be a bit cleaner. What do you think?

cpisciotta commented:
@pepicrft Thanks for your review and the feedback on my questions! Working on a few fixes to your address your feedback and to align with your answers!

cpisciotta commented:
@eliperkins Thanks for the feedback! Yes, that definitely makes sense to me, and I'll update to implement your suggested approach! Originally, I was trying to avoid a larger refactor for something like that, but I agree the switch statements are unideal, and it'll make for a cleaner implementation.

cpisciotta committed:
Delete file header from OutputFormat+ExpressibleByArgument.swift

cpisciotta committed:
Create GitHubRenderer.swift

cpisciotta committed:
Introduce TerminalRenderer

cpisciotta committed:
Rename _colored to colored

cpisciotta committed:
Introduce GitHubRenderer implementation

cpisciotta committed:
Delete unused formatSummary

cpisciotta committed:
Rename OutputFormat to Renderer

cpisciotta committed:
Add default OutputRendering beautify implementation

cpisciotta committed:
Add groups to format methods

cpisciotta committed:
Resolve unit test errors related to outputType

cpisciotta committed:
Temporarily rename to show diff

cpisciotta commented on Sources/XcbeautifyLib/String+Beautify.swift:
Note to self: Rename file to TerminalRenderer.

pepicrft commented on Sources/XcbeautifyLib/Parser.swift:
```diff
@@ -100,13 +100,20 @@ public class Parser {
+            self.renderer = TerminalRenderer(colored: colored)
+        case .gitHubActions:
+            self.renderer = GitHubRenderer()
+        }
```
So much nicer this way :)

cpisciotta committed:
Refactor OutputRendering methods

cpisciotta committed:
Revert colored default value

cpisciotta committed:
Update OutputRendering.swift

cpisciotta committed:
Refactor renderer implementations

- Share default implementation in OutputRendering
- Add initial implementation to GitHubRenderer
- Delete unnecessary implementations in TerminalRenderer

cpisciotta committed:
Add FileComponents

cpisciotta committed:
Refactor using asFileComponents()

cpisciotta committed:
Update TerminalRenderer.swift

cpisciotta committed:
Refactor GitHubRenderer.swift

cpisciotta committed:
Refactor TestSummary

cpisciotta committed:
Refactor outputGitHubActionsLog

cpisciotta committed:
Refactor asFileComponents method

cpisciotta committed:
Prevent colored output when using GitHubRenderer

cpisciotta committed:
Rename standard to terminal

cpisciotta committed:
Delete colored parameter

cpisciotta committed:
Rename GitHubRenderer to GitHubActionsRenderer

cpisciotta committed:
Update GitHubActionsRendererTests.swift

cpisciotta committed:
Update TerminalRendererTests.swift

cpisciotta committed:
Resolve compiler warnings

cpisciotta committed:
Delete unnecessary init

cpisciotta committed:
Remove symbols in GitHubActionsRenderer

cpisciotta committed:
Resolve formatTestCaseMeasured test error

cpisciotta committed:
Extract test methods to renderers

cpisciotta committed:
Resolve clang error

cpisciotta committed:
Resolve testSymbolReferencedFrom

cpisciotta committed:
Update XCTestManifests.swift

cpisciotta committed:
Refactor TestSummary

cpisciotta committed:
Resolve Ubuntu errors

cpisciotta committed:
Update XCTestManifests.swift

cpisciotta commented on Sources/XcbeautifyLib/Parser.swift:
```diff
@@ -67,7 +69,7 @@ public class Parser {
         innerParser(.clangError, outputType: .error),
         innerParser(.checkDependenciesErrors, outputType: .error),
-        innerParser(.provisioningProfileRequired, outputType: .warning),
+        innerParser(.provisioningProfileRequired, outputType: .error),
```
Looks like this wasn't caught in previous releases, but - looking at the existing implementation and tests - I think this and the one below should be an error.

eliperkins replied:
Is this change worth moving to a separate PR, so as not to couple the new functionality of this PR with this (correct) change?

cpisciotta replied:
Good call! I created this PR: https://github.com/tuist/xcbeautify/pull/123.

cpisciotta commented on Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift:
```diff
@@ -0,0 +1,286 @@
+        case error
+    }
+
+    let colored: Bool = false
```
GitHub Actions annotations don't properly render colored output, so I decided to disallow it.

cpisciotta committed:
Extract remaining indent

cpisciotta commented on Sources/XcbeautifyLib/String+Beautify.swift:
I tried to avoid this showing as a deleted file, but I couldn't get it to work on the GH UI. Much of this logic now exists in `TerminalRenderer`.

cpisciotta commented on Tests/XcbeautifyLibTests/RendererTests/GitHubActionsRendererTests.swift:
```diff
@@ -0,0 +1,676 @@
+        XCTAssertEqual(parser.outputType, .task)
+    }
+
+    func testCheckDependenciesErrors() {
+    }
+
+    func testCheckDependencies() {
+    }
```
These empty tests aren't defined for the existing implementation, so totally clear on the expected output.

cpisciotta commented on Tests/XcbeautifyLibTests/XCTestManifests.swift:
There's still some missing tests from this list, but I think it's best to include those in a follow up PR, since many were already missing before I started.

cpisciotta commented:
Hi @pepicrft and @eliperkins. I think I've gotten the PR into a state ready for review and ready for production. There's quite a lot of changes, so I'm happy to walk through any areas where there's concern or questions.

Working on a PR description...

eliperkins commented on Sources/XcbeautifyLib/Parser.swift:
```diff
@@ -1,10 +1,12 @@
     private let additionalLines: () -> String?
 
-    public var summary: TestSummary? = nil
+    private(set) var summary: TestSummary? = nil
```
Since this changes a public `{ get set }` to only a public `{ get }`, is this a breaking change for API consumers? If so, what do we need to do to mark the next version of xcbeautify as a breaking change?

Otherwise, can we revert this change?

cpisciotta replied:
This was addressed in a different PR. I don't think it requires a major release, since the only consumer of `XcbeautifyLib` should be the executable target. @pepicrft should confirm.

eliperkins commented on Sources/XcbeautifyLib/Parser.swift:
```diff
@@ -151,7 +163,12 @@ public class Parser {
+        if let summary = summary {
+            write(summary.format(with: renderer))
+        }
+    }
```
For my understanding, what does this method provide for API consumers that `write(_:)` doesn't?

cpisciotta replied:
Addressed in a different PR and this implementation approach is no longer relevant.

eliperkins commented on Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift:
```diff
@@ -0,0 +1,279 @@
+        let fileComponents = filePath.asFileComponents()
+        let reason = group.reason
+
+        // Read 2 additional lines to get the error line and cursor position
+        let line: String = additionalLines() ?? ""
+        let cursor: String = additionalLines() ?? ""
```
You definitely have a better understanding of this parser than I do at this point, so I'm gonna ask a couple questions to help me better understand:

- Is reading `additionalLines` into a variable a common pattern throughout the codebase already?
- What happens if the format of these lines change within Xcode's output? Are there ways to safeguard against potential future changes?

cpisciotta replied:
> Is reading additionalLines into a variable a common pattern throughout the codebase already?

Yes it is. That's how the pre-existing implementation (now `TerminalRenderer`) is implemented.
 https://github.com/tuist/xcbeautify/blob/a49d6e54189da4f208e61e98bc23c11fd6ab9d0d/Sources/XcbeautifyLib/Renderers/TerminalRenderer.swift#L241-L242

> What happens if the format of these lines change within Xcode's output? Are there ways to safeguard against potential future changes?

There aren't any. We'd need to add additional logic, but I think that's out of the scope of this PR. I'm not sure it's even something worth implementing unless the Xcode output actually changes.

eliperkins commented on Sources/XcbeautifyLib/Renderers/GitHubActionsRenderer.swift:
```diff
@@ -0,0 +1,279 @@
+}
+
+
+private struct FileComponents {
+    private let path: String
+    private let line: Int?
+    private let column: Int?
+
+    init(path: String, line: Int?, column: Int?) {
+        self.path = path
+        self.line = line
+        self.column = column
+    }
+
+    var formatted: String {
+        guard let line = line else {
+            return "file=\(path)"
+        }
+
+        guard let column = column else {
+            return "file=\(path),line=\(line)"
+        }
+
+        return "file=\(path),line=\(line),col=\(column)"
+    }
+}
```
What do you think about pulling this type out into a separate file, even though it's only used here in this class? That might make it easier to find in the future.

cpisciotta replied:
I'd prefer to keep it in this file, since this is specifically designed to only be used with the `GitHubActionsRenderer`.

eliperkins commented on Sources/XcbeautifyLib/Renderers/OutputRendering.swift:
```diff
@@ -0,0 +1,521 @@
+        line: String,
+        pattern: Pattern,
+        additionalLines: @escaping () -> (String?)
+    ) -> String? {
```
This function also has some typecasting checks that might be worth testing. Is the output of this tested within the existing test suite?

cpisciotta replied:
Yes! All of the type casts here should be covered by the existing unit tests. Any failure to type cast will trigger an assertion failure.

eliperkins commented on Sources/XcbeautifyLib/Renderers/OutputRendering.swift:
```diff
@@ -0,0 +1,521 @@
+        case (.aggregateTarget, let group as AggregateTargetCaptureGroup):
+            return formatTargetCommand(command: "Aggregate", group: group)
+        case (.analyze, let group as AnalyzeCaptureGroup):
+            return  formatAnalyze(group: group)
```
```suggestion
            return formatAnalyze(group: group)
```

eliperkins commented on Sources/XcbeautifyLib/Renderers/OutputRendering.swift:
```diff
@@ -0,0 +1,521 @@
+
+    func formatLinking(group: LinkingCaptureGroup) -> String {
+        let target = group.target
+#if os(Linux)
```
Nice! Thanks for handling Linux too :heart:

eliperkins commented on Sources/XcbeautifyLib/String+Colored.swift:
```diff
@@ -0,0 +1,17 @@
+import Foundation
+
+extension String {
+    func coloredTime() -> String {
+        guard let time = Double(self) else { return self }
+        if time < 0.025 { return self }
+        if time < 0.100 { return f.Yellow }
+        return f.Red
+    }
+
+    func coloredDeviation() -> String {
+        guard let deviation = Double(self) else { return self }
+        if deviation < 1 { return self }
+        if deviation < 10 { return f.Yellow }
+        return f.Red
+    }
```
Would these functions be clearer if written as switch statements and (open) ranges?

cpisciotta replied:
This is the existing implementation from `TerminalRenderer`, and - in this specific case - I don't think it'd be any clearer with just a few branches.

eliperkins commented on Tests/XcbeautifyLibTests/RendererTests/GitHubActionsRendererTests.swift:
```diff
@@ -0,0 +1,676 @@
+    }
+
+    func testCompile() {
+#if os(macOS)
```
Is there a Linux branch of this that should be tested as well?

cpisciotta replied:
Since this is copied from `TerminalRendererTests`, I'm not sure what the expected output is. I can look at this in a follow-up PR, since this is an issue for all `OutputRendering` types.

eliperkins commented on Tests/XcbeautifyLibTests/RendererTests/GitHubActionsRendererTests.swift:
```diff
@@ -0,0 +1,676 @@
+        XCTAssertFalse(parser.needToRecordSummary)
+        let formatted1 = logFormatted(input1)
+#if os(macOS)
+        // FIXME: Failing on Linux
```
Is this in scope for this PR?

cpisciotta replied:
I think this should be a separate PR, since this is copied from `TerminalRendererTests` and also an issue there. I can look into why this is triggering an error and fix for all `OutputRendering` types.

eliperkins reviewed (approved):
Great work on this! Admittedly, I'm not an expert at the internals of xcbeautify, so I asked a few questions to help better understand this codebase.

Thanks for improving the ecosystem with this PR! ❤️

cpisciotta committed:
Add doc comment to GitHubActionsRenderer

cpisciotta committed:
Format OutputRendering file

cpisciotta commented:
To better separate the scope of this PR and to improve readability (for review and history), I've introduced one more intermediate PR (https://github.com/tuist/xcbeautify/pull/124). In the new PR, I do all of the refactor work to introduce the concept of a renderer before we define a new one here.

cpisciotta committed:
Fix rebase conflicts

cpisciotta committed:
Refactor TestSummary

cpisciotta commented:
@eliperkins Are you aware of any way to filter annotations on GitHub Actions? I'm noticing a lot of noise in the summary section (namely from dependencies), so I'm trying to decide if it makes sense to add an "ignored file paths" option to xcbeautify. Alternatively, It'd be great if GitHub had this as some type of configuration option.

pepicrft reviewed (approved):
Amazing work on this one @cpisciotta 👏🏼. I tried it locally and got the errors properly formatted:

```bash
::error file=/Users/pepicrft/Downloads/MyApp/MyApp/ContentView.swift,line=18,col=9::expected declaration
        .padding()
        ^
::error file=/Users/pepicrft/Downloads/MyApp/MyApp/ContentView.swift,line=20,col=1::extraneous '}' at top level
}
^
::error file=/Users/pepicrft/Downloads/MyApp/MyApp/ContentView.swift,line=12,col=9::cannot find 'VStac' in scope
        VStac
        ^~~~~
::error file=/Users/pepicrft/Downloads/MyApp/MyApp/ContentView.swift,line=20,col=1::extraneous '}' at top level
}
^
::error file=/Users/pepicrft/Downloads/MyApp/MyApp/ContentView.swift,line=18,col=9::expected declaration
```

pepicrft commented:
@all-contributors add @cpisciotta for code

tonyarnold commented:
I'm not sure if this is directly related to this issue, but I'm seeing truncated results in my actions summary:

```
[Test SDK (tvOS Simulator): --- xcodebuild#L1](somerepolink)
Using the first of multiple matching destinations:
```

Is all I get, when there should be a bunch of lines below this.

cpisciotta commented:
@tonyarnold I think that's unrelated to my changes here. I suspect it fails for the same or a similar reason to https://github.com/tuist/xcbeautify/issues/105. Can you give a little more context about the lines above and below?

tonyarnold commented:
@cpisciotta yeah, it sounds like #105 — this was the full error in my build logs:

```
xcodebuild: error: Unable to find a destination matching the provided destination specifier:
		{ generic:1, platform:visionOS Simulator }

	Available destinations for the "RevealServer" scheme:
		{ platform:macOS, arch:arm64, id:00006000-001A605E3E23801E }
		{ platform:macOS, arch:x86_64, id:00006000-001A605E3E23801E }
		{ platform:macOS, arch:arm64, variant:Mac Catalyst, id:00006000-001A605E3E23801E }
		{ platform:macOS, arch:x86_64, variant:Mac Catalyst, id:00006000-001A605E3E23801E }
		{ platform:macOS, arch:arm64, variant:Designed for [iPad,iPhone], id:00006000-001A605E3E23801E }
		{ platform:watchOS, arch:arm64_32, id:00008006-0013708C36A2002E, name:Tony’s Apple Watch }
		{ platform:iOS, arch:arm64, id:00008120-000E48EC1178C01E, name:Tony’s iPhone }
		{ platform:iOS, id:dvtdevice-DVTiPhonePlaceholder-iphoneos:placeholder, name:Any iOS Device }
		{ platform:iOS Simulator, id:dvtdevice-DVTiOSDeviceSimulatorPlaceholder-iphonesimulator:placeholder, name:Any iOS Simulator Device }
		{ platform:macOS, name:Any Mac }
		{ platform:macOS, variant:Mac Catalyst, name:Any Mac }
		{ platform:tvOS, id:dvtdevice-DVTiOSDevicePlaceholder-appletvos:placeholder, name:Any tvOS Device }
		{ platform:tvOS Simulator, id:dvtdevice-DVTiOSDeviceSimulatorPlaceholder-appletvsimulator:placeholder, name:Any tvOS Simulator Device }
		{ platform:visionOS, variant:Designed for [iPad,iPhone], id:dvtdevice-DVTiOSDevicePlaceholder-xros:placeholder, name:Any visionOS Device }
		{ platform:watchOS, id:dvtdevice-DVTiOSDevicePlaceholder-watchos:placeholder, name:Any watchOS Device }
		{ platform:watchOS Simulator, id:dvtdevice-DVTiOSDeviceSimulatorPlaceholder-watchsimulator:placeholder, name:Any watchOS Simulator Device }
		{ platform:tvOS Simulator, id:DA53B578-CBC9-42FC-8801-30A0B894C730, OS:16.4, name:Apple TV }
		{ platform:tvOS Simulator, id:DA53B578-CBC9-42FC-8801-30A0B894C730, OS:16.4, name:Apple TV }
		{ platform:tvOS Simulator, id:D72208D9-8C11-4CF7-9601-EEE0119CD1B4, OS:17.0, name:Apple TV }
		{ platform:tvOS Simulator, id:D72208D9-8C11-4CF7-9601-EEE0119CD1B4, OS:17.0, name:Apple TV }
		{ platform:tvOS Simulator, id:A1C72F8C-6A78-40FA-88FC-0905C69F3B69, OS:16.4, name:Apple TV 4K (3rd generation) }
		{ platform:tvOS Simulator, id:A1C72F8C-6A78-40FA-88FC-0905C69F3B69, OS:16.4, name:Apple TV 4K (3rd generation) }
		{ platform:tvOS Simulator, id:063F7B75-9A31-4A3F-9526-0399C55FA768, OS:17.0, name:Apple TV 4K (3rd generation) }
		{ platform:tvOS Simulator, id:063F7B75-9A31-4A3F-9526-0399C55FA768, OS:17.0, name:Apple TV 4K (3rd generation) }
		{ platform:tvOS Simulator, id:B2BEA62E-C76E-4FF3-8A25-D5AFD2F86F00, OS:16.4, name:Apple TV 4K (3rd generation) (at 1080p) }
		{ platform:tvOS Simulator, id:B2BEA62E-C76E-4FF3-8A25-D5AFD2F86F00, OS:16.4, name:Apple TV 4K (3rd generation) (at 1080p) }
		{ platform:tvOS Simulator, id:C7C6F39E-F8EB-4E0B-BF90-F943E461B025, OS:17.0, name:Apple TV 4K (3rd generation) (at 1080p) }
		{ platform:tvOS Simulator, id:C7C6F39E-F8EB-4E0B-BF90-F943E461B025, OS:17.0, name:Apple TV 4K (3rd generation) (at 1080p) }
		{ platform:watchOS Simulator, id:148740F0-D03F-459F-B9FC-CEA4CC5B1C6F, OS:9.4, name:Apple Watch SE (40mm) (2nd generation) }
		{ platform:watchOS Simulator, id:148740F0-D03F-459F-B9FC-CEA4CC5B1C6F, OS:9.4, name:Apple Watch SE (40mm) (2nd generation) }
		{ platform:watchOS Simulator, id:E288AAEB-85D4-4D87-A662-9EFC4DBF18D4, OS:10.0, name:Apple Watch SE (40mm) (2nd generation) }
		{ platform:watchOS Simulator, id:E288AAEB-85D4-4D87-A662-9EFC4DBF18D4, OS:10.0, name:Apple Watch SE (40mm) (2nd generation) }
		{ platform:watchOS Simulator, id:05E5301A-B673-448E-B6F9-54738DCD77DC, OS:9.4, name:Apple Watch SE (44mm) (2nd generation) }
		{ platform:watchOS Simulator, id:05E5301A-B673-448E-B6F9-54738DCD77DC, OS:9.4, name:Apple Watch SE (44mm) (2nd generation) }
		{ platform:watchOS Simulator, id:0C5EC9D7-30C2-4DC2-A422-657D502E4B02, OS:10.0, name:Apple Watch SE (44mm) (2nd generation) }
		{ platform:watchOS Simulator, id:0C5EC9D7-30C2-4DC2-A422-657D502E4B02, OS:10.0, name:Apple Watch SE (44mm) (2nd generation) }
		{ platform:watchOS Simulator, id:4799067A-05C2-4BC0-BC4B-E7D9BE61067B, OS:9.4, name:Apple Watch Series 5 (40mm) }
		{ platform:watchOS Simulator, id:4799067A-05C2-4BC0-BC4B-E7D9BE61067B, OS:9.4, name:Apple Watch Series 5 (40mm) }
		{ platform:watchOS Simulator, id:2E187D3C-7CD8-4D05-B1BA-B244212CE509, OS:10.0, name:Apple Watch Series 5 (40mm) }
		{ platform:watchOS Simulator, id:2E187D3C-7CD8-4D05-B1BA-B244212CE509, OS:10.0, name:Apple Watch Series 5 (40mm) }
		{ platform:watchOS Simulator, id:BE34C017-60B3-4C7D-B5ED-D8357013FDEE, OS:9.4, name:Apple Watch Series 5 (44mm) }
		{ platform:watchOS Simulator, id:BE34C017-60B3-4C7D-B5ED-D8357013FDEE, OS:9.4, name:Apple Watch Series 5 (44mm) }
		{ platform:watchOS Simulator, id:DB94E16D-1CB7-4CE0-9BFF-270760EEA49C, OS:10.0, name:Apple Watch Series 5 (44mm) }
		{ platform:watchOS Simulator, id:DB94E16D-1CB7-4CE0-9BFF-270760EEA49C, OS:10.0, name:Apple Watch Series 5 (44mm) }
		{ platform:watchOS Simulator, id:9DBCB88D-6131-4F29-ABA2-FD5E084018F2, OS:9.4, name:Apple Watch Series 6 (40mm) }
		{ platform:watchOS Simulator, id:9DBCB88D-6131-4F29-ABA2-FD5E084018F2, OS:9.4, name:Apple Watch Series 6 (40mm) }
		{ platform:watchOS Simulator, id:283BD7AB-171A-493D-9F5C-BD66689606A6, OS:10.0, name:Apple Watch Series 6 (40mm) }
		{ platform:watchOS Simulator, id:283BD7AB-171A-493D-9F5C-BD66689606A6, OS:10.0, name:Apple Watch Series 6 (40mm) }
		{ platform:watchOS Simulator, id:CCD5EF60-3D40-44FC-90B4-123611312DF8, OS:9.4, name:Apple Watch Series 6 (44mm) }
		{ platform:watchOS Simulator, id:CCD5EF60-3D40-44FC-90B4-123611312DF8, OS:9.4, name:Apple Watch Series 6 (44mm) }
		{ platform:watchOS Simulator, id:5734BF15-99FB-4C57-96B7-6D065C07665D, OS:10.0, name:Apple Watch Series 6 (44mm) }
		{ platform:watchOS Simulator, id:5734BF15-99FB-4C57-96B7-6D065C07665D, OS:10.0, name:Apple Watch Series 6 (44mm) }
		{ platform:watchOS Simulator, id:A5732026-F09D-48D6-BD13-8BA70A03F6B4, OS:9.4, name:Apple Watch Series 7 (41mm) }
		{ platform:watchOS Simulator, id:A5732026-F09D-48D6-BD13-8BA70A03F6B4, OS:9.4, name:Apple Watch Series 7 (41mm) }
		{ platform:watchOS Simulator, id:EBBE8C39-4413-427E-8565-4E6917F57DAF, OS:10.0, name:Apple Watch Series 7 (41mm) }
		{ platform:watchOS Simulator, id:EBBE8C39-4413-427E-8565-4E6917F57DAF, OS:10.0, name:Apple Watch Series 7 (41mm) }
		{ platform:watchOS Simulator, id:6E05738F-ECF0-419E-8242-46F2F99ADEA0, OS:9.4, name:Apple Watch Series 7 (45mm) }
		{ platform:watchOS Simulator, id:6E05738F-ECF0-419E-8242-46F2F99ADEA0, OS:9.4, name:Apple Watch Series 7 (45mm) }
		{ platform:watchOS Simulator, id:8122EFC6-EB9C-486A-A590-4C286ED9A427, OS:10.0, name:Apple Watch Series 7 (45mm) }
		{ platform:watchOS Simulator, id:8122EFC6-EB9C-486A-A590-4C286ED9A427, OS:10.0, name:Apple Watch Series 7 (45mm) }
		{ platform:watchOS Simulator, id:9D5960D0-A4A6-48CD-8578-5E1864360186, OS:9.4, name:Apple Watch Series 8 (41mm) }
		{ platform:watchOS Simulator, id:9D5960D0-A4A6-48CD-8578-5E1864360186, OS:9.4, name:Apple Watch Series 8 (41mm) }
		{ platform:watchOS Simulator, id:3B177FA5-A381-473C-AAE2-445E998F99C4, OS:10.0, name:Apple Watch Series 8 (41mm) }
		{ platform:watchOS Simulator, id:3B177FA5-A381-473C-AAE2-445E998F99C4, OS:10.0, name:Apple Watch Series 8 (41mm) }
		{ platform:watchOS Simulator, id:08EC3202-7548-4B1D-9AF5-A99DADEB7E40, OS:9.4, name:Apple Watch Series 8 (45mm) }
		{ platform:watchOS Simulator, id:08EC3202-7548-4B1D-9AF5-A99DADEB7E40, OS:9.4, name:Apple Watch Series 8 (45mm) }
		{ platform:watchOS Simulator, id:BB5EE960-F1DD-466D-A029-EBA766604BC8, OS:10.0, name:Apple Watch Series 8 (45mm) }
		{ platform:watchOS Simulator, id:BB5EE960-F1DD-466D-A029-EBA766604BC8, OS:10.0, name:Apple Watch Series 8 (45mm) }
		{ platform:watchOS Simulator, id:F0848D83-8E3E-4E74-A0F2-5F1B7CB9C4B7, OS:9.4, name:Apple Watch Ultra (49mm) }
		{ platform:watchOS Simulator, id:F0848D83-8E3E-4E74-A0F2-5F1B7CB9C4B7, OS:9.4, name:Apple Watch Ultra (49mm) }
		{ platform:watchOS Simulator, id:AD7CB514-7034-4D4D-91A4-7B93D4635D3A, OS:10.0, name:Apple Watch Ultra (49mm) }
		{ platform:watchOS Simulator, id:AD7CB514-7034-4D4D-91A4-7B93D4635D3A, OS:10.0, name:Apple Watch Ultra (49mm) }
		{ platform:iOS Simulator, id:186BE321-E30C-4CF5-98B3-AC163EFB6B39, OS:17.0, name:iPad (10th generation) }
		{ platform:iOS Simulator, id:186BE321-E30C-4CF5-98B3-AC163EFB6B39, OS:17.0, name:iPad (10th generation) }
		{ platform:iOS Simulator, id:540D146C-6433-4950-8ACE-7536A74035AC, OS:17.0, name:iPad Air (5th generation) }
		{ platform:iOS Simulator, id:540D146C-6433-4950-8ACE-7536A74035AC, OS:17.0, name:iPad Air (5th generation) }
		{ platform:iOS Simulator, id:80F7F171-00CA-4DB0-86D2-0B36A05B68F1, OS:17.0, name:iPad Pro (11-inch) (4th generation) }
		{ platform:iOS Simulator, id:80F7F171-00CA-4DB0-86D2-0B36A05B68F1, OS:17.0, name:iPad Pro (11-inch) (4th generation) }
		{ platform:iOS Simulator, id:D834BA03-F26F-47FF-A77E-7670BC433A5A, OS:17.0, name:iPad Pro (12.9-inch) (6th generation) }
		{ platform:iOS Simulator, id:D834BA03-F26F-47FF-A77E-7670BC433A5A, OS:17.0, name:iPad Pro (12.9-inch) (6th generation) }
		{ platform:iOS Simulator, id:2FD97D0E-517A-4860-8CB3-17EBB156694F, OS:17.0, name:iPad mini (6th generation) }
		{ platform:iOS Simulator, id:2FD97D0E-517A-4860-8CB3-17EBB156694F, OS:17.0, name:iPad mini (6th generation) }
		{ platform:iOS Simulator, id:D47D0A82-C38A-4A6C-8251-68DDD0B9C916, OS:17.0, name:iPhone 14 }
		{ platform:iOS Simulator, id:D47D0A82-C38A-4A6C-8251-68DDD0B9C916, OS:17.0, name:iPhone 14 }
		{ platform:iOS Simulator, id:84682904-CE87-43A0-B4AA-3A8158D3F8D2, OS:17.0, name:iPhone 14 Plus }
		{ platform:iOS Simulator, id:84682904-CE87-43A0-B4AA-3A8158D3F8D2, OS:17.0, name:iPhone 14 Plus }
		{ platform:iOS Simulator, id:D8A94882-A6AF-426A-8372-717905864D49, OS:17.0, name:iPhone 14 Pro }
		{ platform:iOS Simulator, id:D8A94882-A6AF-426A-8372-717905864D49, OS:17.0, name:iPhone 14 Pro }
		{ platform:iOS Simulator, id:6D839990-6250-4164-8EAB-3D7AD9829B8A, OS:17.0, name:iPhone 14 Pro Max }
		{ platform:iOS Simulator, id:6D839990-6250-4164-8EAB-3D7AD9829B8A, OS:17.0, name:iPhone 14 Pro Max }
		{ platform:iOS Simulator, id:194F66D9-A7A1-4491-8989-3B25FD351D7C, OS:17.0, name:iPhone SE (3rd generation) }
		{ platform:iOS Simulator, id:194F66D9-A7A1-4491-8989-3B25FD351D7C, OS:17.0, name:iPhone SE (3rd generation) }

	Ineligible destinations for the "RevealServer" scheme:
		{ platform:iOS, id:254d9f75b728750efc9b6a75c4c2611ca5a7d97b, name:Ted's iPad, error:To use Ted's iPad for development, enable Developer Mode in Settings → Privacy & Security.  }
```
